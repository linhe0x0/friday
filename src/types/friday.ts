// eslint-disable-next-line no-shadow
export enum EndpointProtocol {
  TCP = 'tcp:',
  HTTP = 'http:',
  UNIX = 'unix:',
}

export interface Endpoint {
  protocol: EndpointProtocol
  host: string
  port?: number
}
