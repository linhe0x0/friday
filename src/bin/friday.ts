#!/usr/bin/env node

import _ from 'lodash'
import yargs from 'yargs'

import { Endpoint, EndpointProtocol } from '../types/friday'
import isValidPort from '../utilities/is-valid-port'
import parseEndpoint from '../utilities/parse-endpoint'
import resolveEntry from '../utilities/resolve-entry'
import serve from '../utilities/serve'

const args = yargs
  .scriptName('friday')
  .option('host', {
    alias: 'h',
    describe: 'specify a host on which to listen',
    type: 'string',
  })
  .option('port', {
    alias: 'p',
    describe: 'specify a port on which to listen',
    type: 'number',
  })
  .option('listen', {
    alias: 'l',
    describe: 'specify a URI endpoint on which to listen',
    type: 'string',
  })
  .option('unix-socket', {
    alias: 'n',
    describe: 'path to a UNIX socket',
    type: 'string',
  })
  .option('silent', {
    alias: 's',
    describe: 'disable all output',
    type: 'boolean',
  })
  .example(
    '',
    `
  For TCP (traditional host/port) endpoint:

    $ friday -l tcp://hostname:1234

  For UNIX domain socket endpoint:

    $ friday -l unix:/path/to/socket.sock
`
  ).argv

const { host, port, listen, silent } = args
const defaultPort = parseInt(process.env.PORT || '3000', 10) || 3000
const defaultHost = '0.0.0.0'

if (_.isNil(process.env.NODE_ENV)) {
  process.env.NODE_ENV = 'production'
}

process.env.FRIDAY_ENV = 'production'

if (silent) {
  process.env.LOGGER_LEVEL = 'silent'
}

const isHostOrPortProvided = !!(host || port)

if (isHostOrPortProvided && listen) {
  console.error('Both host/port and tcp provided. You can only use one.')
  process.exit(1)
}

if (port) {
  if (!isValidPort(port)) {
    console.error(`Port option must be a number. Got: ${port}`)
    process.exit(1)
  }
}

const endpoint: Endpoint = listen
  ? parseEndpoint(listen)
  : {
      protocol: EndpointProtocol.HTTP,
      host: host || defaultHost,
      port,
    }

if (endpoint.protocol !== EndpointProtocol.UNIX) {
  _.defaults(endpoint, {
    host: defaultHost,
    port: defaultPort,
  })
}

const entryFile = resolveEntry(args._[0])

serve(endpoint, entryFile)
  .then(() => {
    let message = 'Server is running.'

    if (endpoint.protocol === EndpointProtocol.UNIX) {
      message = `Server is running at ${endpoint.host}.\n`
    } else {
      message = `Server is running at ${endpoint.protocol}//${endpoint.host}:${endpoint.port}.\n`
    }

    process.stdout.write(message)
  })
  .catch((err) => {
    console.error(err)
    process.exit(1)
  })
