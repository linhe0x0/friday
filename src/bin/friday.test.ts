const runCommand = (...args: string[]): void => {
  process.argv = ['node', 'friday', ...args]

  require('./friday')
}

beforeEach(() => {
  jest.resetModules()
})

afterEach(() => {
  jest.resetAllMocks()
})

describe('friday --help', () => {
  test('should output help info with example message when called with --help flag.', () => {
    const log = jest.spyOn(console, 'log').mockImplementation(() => {})
    const exit = jest
      .spyOn<any, string>(process, 'exit')
      .mockImplementation(() => {})

    runCommand('--help')

    expect(log.mock.calls[0]![0]).toMatchSnapshot()
    expect(exit).toBeCalledWith(0)

    log.mockRestore()
    exit.mockRestore()
  })

  test('should output help info of start command when called with --help flag.', () => {
    const log = jest.spyOn(console, 'log').mockImplementation(() => {})
    const exit = jest
      .spyOn<any, string>(process, 'exit')
      .mockImplementation(() => {})

    runCommand('start', '--help')

    expect(log.mock.calls[0]![0]).toMatchSnapshot()
    expect(exit).toBeCalledWith(0)

    log.mockRestore()
    exit.mockRestore()
  })
})

// Use the following to make the file a module, because the file cannot be
// complied under '--isolatedModules' flag without it.
// eslint-disable-next-line jest/no-export
export {}
