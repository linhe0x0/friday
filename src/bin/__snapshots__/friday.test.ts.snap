// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`friday --help should output help info of start command when called with --help flag. 1`] = `
"friday start [entry-point]

start a server, as it is the default command

Options:
      --version  Show version number                                   [boolean]
      --help     Show help                                             [boolean]
  -h, --host     specify a host on which to listen                      [string]
  -p, --port     specify a port on which to listen                      [number]
  -l, --listen   specify a URI endpoint on which to listen              [string]
  -e, --env      specify an environment of application and config files [string]"
`;

exports[`friday --help should output help info with example message when called with --help flag. 1`] = `
"friday <cmd> [args]

Options:
      --version  Show version number                                   [boolean]
      --help     Show help                                             [boolean]
  -h, --host     specify a host on which to listen                      [string]
  -p, --port     specify a port on which to listen                      [number]
  -l, --listen   specify a URI endpoint on which to listen              [string]
  -e, --env      specify an environment of application and config files [string]

Examples:

    By default, app will be listened on 0.0.0.0:3000 and look first for the
    \\"main\\" field in package.json and subsequently for app.js as the default
    entry_point.

    Listen endpoints (specified by the --listen or -l options above) instruct
    app to listen on the specified interface/port, UNIX domain socket.

    For TCP (traditional host/port) endpoint:

    $ friday -p 1234
    $ friday -h 127.0.0.1
    $ friday -l tcp://127.0.0.1:1234

    For UNIX domain socket endpoint:

    $ friday -l unix:/path/to/socket.sock"
`;
